### ISSUE TITLE
EC2 module results data structures frustratingly inconsistent for idempotent provisioning

### ISSUE THREAD

**Issue by ches** [ansible/ansible#8986]
##### Issue Type:

Bug Report/Feature Idea
##### Ansible Version:

1.7.1
##### Environment:

N/A
##### Summary:

The TL;DR is that the data structures for results from the EC2 module are a bit inconsistent, in a way that hurts the utility of the awesomely handy `count_tag` feature added in 1.5 for idempotent provisioning. This is especially true when you run the module in a loop, which I often do for either `with_sequence` to apply IDs/names, or `with_items` where for example the items are availability zones to provision a class of system in two zones for failover.

That's the summary, but I'd like to give a full example below to make the point, because it shows some cool features of Ansible not quite able to work together. So apologies that that won't be brief :weary:

If changing this behavior is something you're open to, I will take a stab at figuring it out for a pull request.
##### Steps To Reproduce:

Run a play creating instances with the EC2 module and register the result. Output it with `debug`. Then run it again using `count_tag` to get idempotent provisioning. Full example below.
##### Expected Results:

In the normal case (not running module in a loop), you'll get `ec2.instances` as an array of hashes on the first run. On the second run when nothing is created, I would expect an empty array. It is instead an empty _string_. This is passable because at least it'll still pass an emptiness check, but it's unexpected.

The loop case is worse. On first run, `ec2.results.N.instances` are arrays of hashes as expected. On the next run, however, the values of those keys are `null`! I'll demonstrate why this sucks.
##### Actual Results:

Here is an instance launch playbook, a dead simple example of the module, except that we want the playbook to be idempotent with `count_tag`, so a following play is conditional on some instances actually being created. This works fine.

``` yaml

---
- name: Testing results data structure of ec2 module
  hosts: localhost
  connection: local
  gather_facts: false

  tasks:
    - name: Launch an instance
      register: ec2
      ec2:
      args:
        region: ap-southeast-1
        image: ami-12356d40
        instance_type: m3.medium
        key_name: testing
        wait: true
        group: testing
        instance_tags:
          Name: ThrowawayTest
          UniqueID: testing

        # Check/ensure that just this machine is running
        exact_count: 1
        count_tag:
          UniqueID: testing

    - debug: var=ec2.instances

    - name: Add new instances to dynamic host group for deployment
      add_host: name={{ item.public_ip }} groups=launched
      with_items: ec2.instances
      when: item | length > 0
```

Here's the result of a first run:

```
$ ansible-playbook --private-key $AWS_IDENTITY_FILE playbooks/test.yml

PLAY [Testing results data structure of ec2 module] ***************************

TASK: [Launch an instance] ****************************************************
changed: [127.0.0.1 -> 127.0.0.1]

TASK: [debug var=ec2.instances] ***********************************************
ok: [127.0.0.1] => {
    "ec2.instances": [
        {
            "ami_launch_index": "0",
            "SNIPPED"
        }
    ]
}

TASK: [Add new instances to dynamic host group for deployment] ****************
ok: [127.0.0.1] => SNIP

PLAY RECAP ********************************************************************
127.0.0.1                  : ok=3    changed=1    unreachable=0    failed=0
```

And here's the next run. Note the empty string value from the registered var, though the conditional play still skips as desired, incidentally.

```
$ ansible-playbook --private-key $AWS_IDENTITY_FILE playbooks/test.yml

PLAY [Testing results data structure of ec2 module] ***************************

TASK: [Launch an instance] ****************************************************
ok: [127.0.0.1 -> 127.0.0.1]

TASK: [debug var=ec2.instances] ***********************************************
ok: [127.0.0.1] => {
    "ec2.instances": ""
}

TASK: [Add new instances to dynamic host group for deployment] ****************
skipping: [127.0.0.1]

PLAY RECAP ********************************************************************
127.0.0.1                  : ok=3    changed=0    unreachable=0    failed=0
```

Now add a loop for giving instances some sequential attribute identifier. `with_subelements` is beautiful for this pattern.

``` yaml

---

- name: Testing results data structure of ec2 module
  hosts: localhost
  connection: local
  gather_facts: false

  tasks:
    - name: Launch instances
      with_sequence: count=2
      register: ec2
      ec2:
      args:
        region: ap-southeast-1
        image: ami-12356d40
        instance_type: m3.medium
        key_name: testing
        wait: true
        group: testing
        instance_tags:
          Name: ThrowawayTest{{ item }}
          UniqueID: '{{ item }}'

        # Check/ensure that just these machines are running
        exact_count: 1
        count_tag:
          UniqueID: '{{ item }}'

    - debug: var=ec2.results

    - name: Add new instances to dynamic host group for deployment
      add_host: name={{ item.1.public_ip }} groups=launched
      with_subelements:
        - ec2.results
        - instances
      when: item.1 | length > 0
```

The result when there are changes is unsurprising, but for sake of showing the shape of the data structure, here's a trimmed version:

```
$ ansible-playbook --private-key $AWS_IDENTITY_FILE playbooks/test.yml

PLAY [Testing results data structure of ec2 module] ***************************

TASK: [Launch instances] ******************************************************
changed: [127.0.0.1 -> 127.0.0.1] => (item=1)
changed: [127.0.0.1 -> 127.0.0.1] => (item=2)

TASK: [debug var=ec2.results] *************************************************
ok: [127.0.0.1] => {
    "ec2.results": [
        {
            "changed": true,
            "instance_ids": [
                "i-0850b623"
            ],
            "instances": [
                {
                    "ami_launch_index": "0",
                    "SNIPPED"
                }
            ],
            "invocation": {
                "module_args": "",
                "module_name": "ec2"
            },
            "item": "1",
            "tagged_instances": [
                {
                    "ami_launch_index": "0",
                    "SNIPPED"
                }
            ]
        },
        {
            "changed": true,
            "instance_ids": [
                "i-40a3426b"
            ],
            "instances": [
                {
                    "ami_launch_index": "0",
                    "SNIPPED"
                }
            ],
            "invocation": {
                "module_args": "",
                "module_name": "ec2"
            },
            "item": "2",
            "tagged_instances": [
                {
                    "ami_launch_index": "0",
                    "SNIPPED"
                }
            ]
        }
    ]
}

TASK: [Add new instances to dynamic host group for deployment] ****************
ok: [127.0.0.1] => SNIP
ok: [127.0.0.1] => SNIP

PLAY RECAP ********************************************************************
127.0.0.1                  : ok=3    changed=1    unreachable=0    failed=0
```

Finally, the subsequent run in this case is where the lovely `with_subelements` gets foiled and our nice idempotent provisioning playbook breaks. Note the null values.

```
$ ansible-playbook --private-key $AWS_IDENTITY_FILE playbooks/test.yml

PLAY [Testing results data structure of ec2 module] ***************************

TASK: [Launch instances] ******************************************************
ok: [127.0.0.1 -> 127.0.0.1] => (item=1)
ok: [127.0.0.1 -> 127.0.0.1] => (item=2)

TASK: [debug var=ec2.results] *************************************************
ok: [127.0.0.1] => {
    "ec2.results": [
        {
            "changed": false,
            "instance_ids": null,
            "instances": null,
            "invocation": {
                "module_args": "",
                "module_name": "ec2"
            },
            "item": "1",
            "tagged_instances": [
                {
                    "ami_launch_index": "0",
                    "SNIPPED"
                }
            ]
        },
        {
            "changed": false,
            "instance_ids": null,
            "instances": null,
            "invocation": {
                "module_args": "",
                "module_name": "ec2"
            },
            "item": "2",
            "tagged_instances": [
                {
                    "ami_launch_index": "0",
                    "SNIPPED"
                }
            ]
        }
    ]
}

TASK: [Add new instances to dynamic host group for deployment] ****************
fatal: [127.0.0.1] => the key instances should point to a list, got 'None'

FATAL: all hosts have already failed -- aborting

PLAY RECAP ********************************************************************
           to retry, use: --limit @/Users/ches/test.retry

127.0.0.1                  : ok=2    changed=0    unreachable=0    failed=1
```

Thanks for reading all this :smile:

**Comment by jimi-c** [comment]
Hi, thanks for taking the time to report this!

There has actually been some discussion recently about trying standardize the output of cloud modules, without breaking the existing data structures so as to not break existing playbooks that depend on them. We'll definitely be looking at doing this in the future.

**Comment by ches** [comment]
Thanks @jimi-c, [I'm one of those](https://groups.google.com/d/msg/ansible-project/y41lCYt0cXA/I7vavSEkuUoJ) raising discussion I think :wink:  If there are some other threads you're thinking of please let me know!

**Comment by jimi-c** [comment]
@ches the discussion I was thinking actually came up around another PR that wanted to allow modules to execute add_host directly to remove a step, but we thought it might be a better idea to standardize the output of these modules so that add_host could simply take the registered result as a param instead.

**Comment by mpdehaan** [comment]
Hi!

Thanks very much for your interest in Ansible.  It sincerely means a lot to us. 

On September 26, 2014, due to enormous levels of contribution to the project Ansible decided to reorganize module repos, making it easier
for developers to work on the project and for us to more easily manage new contributions and tickets.

We split modules from the main project off into two repos, http://github.com/ansible/ansible-modules-core and http://github.com/ansible/ansible-modules-extras

If you would still like this ticket attended to, we will need your help in having it reopened in one of the two new repos, and instructions are provided below.

We apologize that we are not able to make this transition happen seamlessly, though this is a one-time change and your help is greatly appreciated -- 
this will greatly improve velocity going forward.

Both sets of modules will ship with Ansible, though they'll receive slightly different ticket handling.

To locate where a module lives between 'core' and 'extras'
- Find the module at http://docs.ansible.com/list_of_all_modules.html
- Open the documentation page for that module
- If the bottom of the docs say "This is an extras module", submit your ticket to https://github.com/ansible/ansible-modules-extras
- Otherwise, submit your pull request to update the existing module to https://github.com/ansible/ansible-modules-core
- action_plugins (modules with server side components) still live in the main repo.  If your ticket affects both, open the ticket
   on the module repo just the same.

Additionally, should you need more help with this, you can ask questions on:
- the ansible-project mailing list: https://groups.google.com/forum/#!forum/ansible-project

Thank you very much!
